### DISCUSSION TITLE
[@types/react] `<Form onSubmit>` should allow an async event handler

### DISCUSSION THREAD

**Discussion by jasikpark** [DefinitelyTyped/DefinitelyTyped#66505] [Issues with a @types package] [answered]
React allows passing an async event handler for `onSubmit`, for example, `react-hook-form` constructs [`handleSubmit`](https://react-hook-form.com/docs/useform/handlesubmit), which returns a promise.

`[@types/react]` should allow for an async function to be passed, so that a rule like https://typescript-eslint.io/rules/no-misused-promises/ is not triggered, warning about dangling promises.

While `() => void` can accept `() => Promise<void>` as a subtype, it is useful to explicitly mention that an event handler may be asynchronous

TLDR: [@types/react] should allow async event handlers

[a comment by an automated account is not shown]

**Comment by karlhorky** [comment]
@eps1lon what are your thoughts here on changing the React types to also accept async functions for event handlers?

Here's an example of an error from the recommended configuration of [`@typescript-eslint/no-misused-promises`](https://typescript-eslint.io/rules/no-misused-promises/):

[`typescript-eslint` Playground](https://typescript-eslint.io/play/#ts=5.5.3&fileType=.tsx&code=JYWwDg9gTgLgBAcigUwIYGMYINwChfIAekscAJsgGaoCuANvJTQHabATNwDCE4HyzGAAoAlHADeuOHBQwaUTgB4ARjRgwOcDlzrB0AawC841AGcAnqzii4hgHxwADAF84Aejt5n%2BIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

`index.tsx`

```ts
import 'react';

export default function Component() {
  return <button onClick={async () => 0} />;
}
```

Error message:

```
Promise-returning function provided to attribute where a void return was expected. eslint (@typescript-eslint/no-misused-promises)
```

![Screenshot 2024-08-23 at 16 51 29](https://github.com/user-attachments/assets/a60ffa90-fee6-4fd8-9e5b-9957edab7901)

### Workaround

Disable the lint warnings/errors about promise misuse reported using `checksVoidReturn.attributes = false`:

`eslint.config.ts`

```ts
'@typescript-eslint/no-misused-promises': [
  'error',
  {
    checksVoidReturn: {
      // Allow async functions passed to event handler props
      // - https://github.com/typescript-eslint/typescript-eslint/pull/4623
      // - https://github.com/typescript-eslint/typescript-eslint/issues/4619
      //
      // Although technically, async functions should not
      // be passed to event handler props:
      // - https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66505#discussioncomment-10411110
      attributes: false,
    },
  },
],
```

**Reply by eps1lon** [reply]
They probably shouldn't allow `async` just like Effects don't allow `async` functions. The priority and batching semantics change after the `await` so you should be explicitly using `.then` instead. We didn't allow `async` functions in `startTransition` for the same reason until we added first-class support. 

React won't `await` the event handler so it really doesn't accept a Promise. You shouldn't pass `async` functions to anything accepting synchronous functions.

**Reply by karlhorky** [reply]
Thanks for the answer!

**Reply by karlhorky** [reply]
I guess at some point, it would be cool to see the React team's suggestion of idiomatic, compact async event handler code which uses `async/await` instead of `.then()`

Since big community packages like `react-hook-form` are using `async/await` and passing promises to event handler props (not to mention countless community blog posts), I'm thinking this pattern is pretty widespread and is superficially visually nicer + more ergonomic than the alternative of either promise chaining using `.then()` (which increases nesting and indentation) or using some instantly-invoked async function. But of course, as Sebastian mentioned, it actually represents using event handler values incorrectly.

It's interesting that the React docs (at least [the Responding to Events page](https://19.react.dev/learn/responding-to-events)) also don't include an example of some asynchronous work being performed.

cc @rickhanlonii @mattcarrollcode

**Reply by eps1lon** [reply]
We're discussing this internally. One thing to note is that calling any imperative method on `event` after the `await` would be a noop hinting at `async` event handlers being a footgun. For example, `event.stopPropagation` or `event.preventDefault` would not work after `await`.

The same applies to using `.then` but generally, that always has been the escape hatch when you pass an `async` function to something that treats the function as sync.

We're leaning towards allowing it but no final decision yet. If somebody could start filing a PR with the necessary changes to check ecosystem interop, that'd be helpful.

**Reply by karlhorky** [reply]
Oh nice, thanks for the follow-up!

I'm assuming that the PR has to do with changing the `: void` here on line 2344 of `types/react/index.d.ts` to `: void | Promise<void>`

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/414f1e14897c7bf5ac307abf06b99dfcbe2767b7/types/react/index.d.ts#L2340-L2344

Seems like maybe a `good-first-issue` or `beginner` PR (at least for this first part)

**Reply by ProchaLu** [reply]
@karlhorky @eps1lon Created a small PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70086

**Reply by eps1lon** [reply, marked as answer]
We've discussed this and we don't want to allow async functions at the type level. We're probably going to start warning at runtime when you return a Promise from an event handler. 

We didn't used to warn because we didn't have a good alternative. With React 19 we have: Use an Action or wrap the async function in `startTransition`. 

We may start automatically wrapping async functions in `startTransition`. But with any additive API this comes with the risk of breaking existing usages that relied on existing behavior even though it was never supported. So by officially allowing them now (even though it "just works"), we increase the risk of not being able to make that change because too much of the ecosystem relied on this behavior.

As an escape hatch, you can always wrap the `async` function in an IIFE or use `promise.then()`. This makes it more obvious that something after the Promise resolved is outside of React's control.

**Reply by karlhorky** [reply]
Interesting, thanks for the feedback!

So then it sounds like it's a pretty official React team recommendation to avoid passing async functions to event handler props 👀

I guess that lots of blog posts, videos and other learning resources should consider adapting to this change in the mid-term. 

Even libraries which promote this pattern (albeit in a component instead of an HTML element event handler prop directly) - such as [React Hook Form](https://github.com/react-hook-form/react-hook-form) (cc @bluebill1049) may want to consider what this means for them.

**Reply by jasikpark** [reply]
It really would be wonderful if TypeScript could model async functions as being separate from `() => void` at the type level, rather than requiring lints and whatnot. Unhandled promises are pretty easy to make use of since they fit inside the supertype of `() => void` v.s. `() => Promise<void>`

**Reply by markerikson** [reply]
Frankly that feels like it's going to end up adding another layer of boilerplate for a set of fairly common usage patterns.

Take this example:

```ts
const onClick = async () => {
  await dispatch(someReduxAsyncThunk()).unwrap();
  navigate("/");
}
```

You're saying React would start warning on that?  It's annoying enough to have to wrap async calls in IIFEs inside of `useEffect` as it is.

I understand the idea that "if you access `event` after an `await` it will break", but I'm not sure that is enough of a footgun to justify forbidding async event handlers.  Like, if _that_ is the concern, maybe instrument event objects in dev to warn about accessing those fields instead?

**Reply by vcarl** [reply]
I do note that the patterns currently suggest in [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect#sending-a-post-request) to my eye look very much like recommending async behaviors be done in event handlers; this code (in practice) would require an `await` in the majority of cases I've experienced:
```js 
function handleSubmit(e) {
  e.preventDefault();
  // ✅ Good: Event-specific logic is in the event handler
  post('/api/register', { firstName, lastName });
}
```
Disallowing promise return types doesn't prevent someone from using promises within the handler, which would still run the same risks of stale values e.g.
```js
function handleSubmit(e) {
  post('/api/register', { firstName, lastName })
    .then((response) => { /* … */ });
}
```

**Reply by eps1lon** [reply]
Stale values isn't the concern here but missing batching and imperative methods on `event` having no effect. The linked docs do not recommend using `async` event handlers.

**Reply by vcarl** [reply]
I recognize that but they do recommend doing asychronous behaviors (e.g. `post()`), and I think it's going to have been a common intuition that functions expected to perform asynchronous behaviors are fair game to mark `async`

**Reply by snewell92** [reply]
Async workflows outlive events, so it makes total sense that this would be disallowed at the type level. Re about "functions expected to perform async behaviours" - I think it's more appropriate to say Event Handlers can trigger async workflows not actually perform them. React & the DOM ignore event handlers return values (outside the quirky return false that has been replaced by DOM APIs) and if you try to use callbacks or promises, your code outlives the event, so it is better to extract the actual work out and respect the eventing signal model that the browser/react use (see MDN & React docs).

```ts
// collate all async work. No reference to DOM / Events
async function processSubmission(data: Record<string, string>) {
  try {
    const validatedData = await validateData(data);
    const submission = await postSubmission(validatedData);
    toastUser();
    resetForm();
  } catch(ex) {
    handleError(ex);
  }
}

// keep event handler sync, fire off async work
function handleSubmit(event: FormEvent<HTMLFormElement>) {
  event.preventDefault();

  const data = formDataToJson(event.currentTarget);
  processSubmission(data);
}
```

Operationalizing this in a linter that would keep track of the use of `event`, ensuring it does not escape its scope (akin to ownership?).

Or even better, restricting all event handlers to be sync void functions. 🤓

**Reply by karlhorky** [reply]
> restricting all event handlers to be sync void functions. 🤓

If you want errors / warnings to be reported on async functions passed to event handler props, [use the recommended config of `@typescript-eslint/no-misused-promises`](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66505#discussioncomment-10066351)

**Reply by karlhorky** [reply]
Interesting, looking into it, this `@typescript-eslint/no-misused-promises` rule was almost enabled by default in `create-next-app` in one of the last PRs (June 2024) by @JoshuaKGoldberg (it's part of the `@typescript-eslint/recommended-type-checked` preset):

- https://github.com/vercel/next.js/pull/52845#issuecomment-2090036184

In the comment linked above, it was decided to not use the type checked lint rules because of performance reasons.

So @eps1lon, in case the `@typescript-eslint/recommended-type-checked` preset is re-PRed to `create-next-app` and approved, these warnings / errors (on async functions passed to event handler props like `onClick`, `onSubmit`, etc) will be coming sooner rather than later (at least for Next.js users).

Originally discussion here:

- https://github.com/vercel/next.js/discussions/37151

**Reply by rickhanlonii** [reply]
@markerikson you example has a bug. if the button is clicked twice and there's any kind of delay between them you'll navigate twice. or if you were navigating to different routes, you change routes to back to the first route click. you'd need to either prevent the button from being clicked multiple times with a pending state, or cancel previous promises on each click to do that safely. so it's not just the accessing the event thing.

actions can help solve this, by letting react know about the async work. acitons include a pending state so you can easily disable the button, or if you used useActionState, ordering is handled for you and you can handle multiple clicks in order.

we would not add a warning until actions are the standard and there's buy in that they are a better DX and remove these types of footguns.

**Reply by silvenon** [reply]
I wanted to pitch in and say that I appreciate the concern for developers by disallowing async event handlers, but I believe that React would be overstepping here a bit. I don't even remember the last time I've written `event.preventDefault()`, and perhaps the warning for an async handler would even be doing me a disservice by not educating me.

TypeScript allows me to make this mistake in vanilla JS:

```ts
deleteBtn.addEventListener('click', async (event) => {
  await deleteItem(event.currentTarget.value)
  event.preventDefault()
})
```

and this way I'm going to learn that it doesn't work and why, and it's so obvious why it doesn't. It's one of those mistakes that I make once and never again, is it really worth it to build an arsenal of disruptive changes around preventing it?

**Reply by karlhorky** [reply]
@silvenon if your main point is about the risk that the warning wouldn't educate developers to the background, would it be an alternative to make any warning experience helpful and educational?

Eg. adding informational error messages but also with links to error pages explaining the failure cases?

From your example above, it sounds as if the alternative is for developers to learn by themselves without tooling (in this case a React warning) providing a helpful hint?

**Reply by silvenon** [reply]
I guess my point is that people who don't know that a delayed `event.preventDefault()` won't work are likely to make that mistake either way:

```tsx
onClick={(event) => {
  deleteItem(event.currentTarget.value).then(() => {
    event.preventDefault()
  })
}}
```

But if they try to add an async handler and get the warning with the explanation link, ok, so now they know why, but they still can't use the `async` handler, and after a while I feel like they're almost as likely to call it delayed in a sync as in an `async` handler.

Or do you think the sync handler be an efficient deterrent because `async/await` makes things seem so simple? As if, people who called it delayed may not have known they did that?

**Reply by karlhorky** [reply]
Right ok, so if I'm understanding correctly you mean that an async function warning will not prevent all permutations of potential asynchronous bug patterns - I can agree with that. There will continue to be other ways that race conditions and other async bugs can be introduced. 

I guess I see the warning about the async handler as is a reduction of the total possible bug patterns, by targeting and educating on a common pattern that people choose.

And with the possibility of suggesting wrapping in `startTransition` instead and keeping it async, it's less likely people will choose the sync version and more likely they will avoid the buggy pattern - the pit of success grows larger.

**Reply by jasikpark** [reply]
The fact that promises run on creation makes the difference between an async and sync event handler much less obvious to me

**Reply by giuseppeg** [reply]
I often use async functions in event handlers. 

Using an IIFE or `.then` won't solve the problem IMHO because folks who will want to keep the async event handler function will just go for it likely without using `useTransition` because not fully aware. The result would be to make this painful but not as effective.

Unless this change comes in a major version, won't it be better to:

1. warn when event handlers return promises (until next major version)
2. introduce an optional provider (like `<StrictMode>`) that, when present, will signal React to automatically wrap those event handlers
